### ISSUE TITLE
The JSON RPC codec is touchy.

### ISSUE THREAD

**Issue by cozylife** [ethereum/go-ethereum#1709]
When sending JSON data through the JSON RPC, it must be a single-line JSON string, and jQuery doesn't parse the JSON object into a proper string if you pass it the object version of the JSON. Please make your codec less touchy about the string formatting or instruct developers on how the JSON string must be formatted when interacting with the JSON RPC.

**Comment by karalabe** [comment]
Can you give a concrete example that we can test against? AFAIK it shouldn't matter how many lines the JSON is made up of. If it does, that it most certainly is a bug that should be fixed.

Formatting wise, Go is very strict is parsing JSON exactly how it was speced. I see this as a virtue as it forces people to send valid data and not accept hackish solutions. But again, I don't know what's failing to it's hard to debate.

**Comment by cozylife** [comment]
I agree that the more strict software is will improve the software by reducing bugs, however, it's not a formatting issue that's causing that error. Try passing a valid object to jQuery's AJAX functionality and it will not work.

```
$.ajax({
	method: 'POST',
	url: 'http://' + jsonSettings['rpcHost'] + ':' + jsonSettings['rpcPort'],
	data: {
			"jsonrpc":"2.0",
			"method":"personal_newAccount",
			"params":["password here"],
			"id":1
		}
}).done(function(response) {
	console.log(response);
});
```

JSON.stringify will correctly parse the object into a usable string, though if I write the string myself as a multiline object, then it'll fail to parse. (Just make the above JSON object a string by wrapping it in quotes and passing it to AJAX.

Also, there never seems to be any mention of what the "id" parameter does, so I just left it at 1. What does it do? The [JSON RPC specifications](http://www.jsonrpc.org/specification#response_object) seem to treat it as an internal use item. (Such as a way to identify the transaction with the server if the client sends multiple transmissions asynchronously.) I've never used JSON RPC before, though I know AJAX. Am I correct in my assumption?

**Comment by tgerring** [comment]
@cozylife Would it be possible to provide an example invoked via cURL that demonstrates the functionality broken versus working? I'm not sure I fully understand the problem quite yet.

The "id" field is indeed an identifier for the request, for example to differentiate calls in batch requests. But there's also nuanced functionality when the id is missing, the request is considered "notification only" and not in need of response form the server. See http://www.jsonrpc.org/specification#notification for more info

**Comment by cozylife** [comment]
I haven't tried it in CURL. I'd imagine it will be the same as with AJAX. You just wrap the object in a string. I'm not sure if it's getting thrown off with my using tabs instead of spaces or if it really just doesn't like multiline JSON.

**Comment by karalabe** [comment]
The idea is that unless you give us something we can just type and it magically runs/fails, it's hard to reproduce. We don't have jquery set up, we don't work with html/js/jquery to easily just paset something like that in and have it run. Also saying to "just wrap the object in a string" is very non obvious if you're used to a different programming envrionment/world/concepts. This is why @tgerring was asking for a cURL example that we can paste into a console and jsut see what the issue is.

If you want to go down the HTML/jQuery path, that's also ok, but then please write up a full working example (all boilerplate included), and also one where the thing fails. I don't want to start reading up on jQuery and waste time assembling an HTML file just to test an RPC call.

**Comment by cozylife** [comment]
```
<html>
<body>
<script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/jquery/2.1.4/jquery.min.js"></script type="text/javascript">
<script>
$.ajax({
    method: 'POST',
    url: 'http://localhost:8545',
    data: {
            "jsonrpc":"2.0",
            "method":"personal_newAccount",
            "params":["password here"],
            "id":1
        }
}).done(function(response) {
    console.log(response);
});
</script>
</body>
</html>
```

**Comment by maran** [comment]
What I suggest is that if you opt-in not to use web3 as your javascript wrapper that you use `JSON.stringify()` on your object. That should solve your issues I think.

If not let me know :)

**Comment by cozylife** [comment]
Yes, I used stringify, and yes it works, though it shouldn't be necessary. Anyway, Web3 doesn't create accounts or unblock them,  and that's why I was using RPC before I switched to IPC.

**Comment by obscuren** [comment]
If the issue still persists please open a new issue.
